In [1]:
import pandas as pd 
import numpy as np 

### Data Paths

In [9]:
grads = "gradrates.xlsx"
acctblty = "Accountability.xlsx"
percapita = "DOR_Income_EQV_Per_Capita.xlsx"
perpupil = "PerPupilExpenditures.xlsx" 
population = "selectedpopulations.xlsx" 
teacher = "teacherdata.xlsx" 

### Datasets

In [12]:
df_grads = pd.read_excel(grads, header = 1)
df_acctblty = pd.read_excel(acctblty, header = 1)
df_percapita = pd.read_excel(percapita, header = 0)
df_perpupil = pd.read_excel(perpupil, header = 1)
df_population = pd.read_excel(population, header = 1)
df_teacher = pd.read_excel(teacher, header = 1)

In [14]:
df_grads

District Name  District Code  \
0         Abby Kelley Foster Charter Public (District)        4450000   
1                                             Abington          10000   
2    Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                     Acton-Boxborough        6000000   
4    Advanced Math and Science Academy Charter (Dis...        4300000   
..                                                 ...            ...   
301                                         Winchester        3440000   
302                                           Winthrop        3460000   
303                                             Woburn        3470000   
304                                          Worcester        3480000   
305                                       State Totals              0   

    # in Cohort  % Graduated  % Still in School  % Non-Grad Completers  \
0            88        100.0                0.0                    0.0   
1           153         88.9                1.3                    0.7   
2            54         96.3                1.9                    0.0   
3           466         96.1                3.2                    0.2   
4           139         99.3                0.7                    0.0   
..          ...          ...                ...                    ...   
301         354         94.1                3.4                    0.3   
302         136         92.6                1.5                    0.0   
303         291         89.7                4.1                    0.3   
304       1,862         85.2                6.1                    1.1   
305      72,602         89.2                4.5                    0.9   

     % H.S. Equiv.  % Dropped Out  % Permanently Excluded  
0              0.0            0.0                     0.0  
1              0.0            9.2                     0.0  
2              0.0            1.9                     0.0  
3              0.0            0.4                     0.0  
4              0.0            0.0                     0.0  
..             ...            ...                     ...  
301            0.3            2.0                     0.0  
302            0.0            5.9                     0.0  
303            0.0            5.8                     0.0  
304            0.1            7.4                     0.0  
305            0.2            5.3                     0.0  

[306 rows x 9 columns]

In [16]:
df_acctblty

District Name  District Code  \
0         Abby Kelley Foster Charter Public (District)        4450000   
1                                             Abington          10000   
2    Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                     Acton-Boxborough        6000000   
4                                             Acushnet          30000   
..                                                 ...            ...   
393                                           Winthrop        3460000   
394                                             Woburn        3470000   
395                                          Worcester        3480000   
396                                        Worthington        3490000   
397                                           Wrentham        3500000   

                       Overall Classification  \
0    Not requiring assistance or intervention   
1    Not requiring assistance or intervention   
2    Not requiring assistance or intervention   
3    Not requiring assistance or intervention   
4    Not requiring assistance or intervention   
..                                        ...   
393  Not requiring assistance or intervention   
394  Not requiring assistance or intervention   
395  Not requiring assistance or intervention   
396  Not requiring assistance or intervention   
397  Not requiring assistance or intervention   

                 Reason for Classification  \
0             Meeting or exceeding targets   
1         Moderate progress toward targets   
2         Moderate progress toward targets   
3             Meeting or exceeding targets   
4      Substantial progress toward targets   
..                                     ...   
393    Substantial progress toward targets   
394       Moderate progress toward targets   
395       Moderate progress toward targets   
396  Limited or no progress toward targets   
397    Substantial progress toward targets   

     Progress Toward Improvement Targets (%)  
0                                       82.0  
1                                       39.0  
2                                       48.0  
3                                       85.0  
4                                       58.0  
..                                       ...  
393                                     56.0  
394                                     32.0  
395                                     41.0  
396                                      9.0  
397                                     63.0  

[398 rows x 5 columns]

In [18]:
df_percapita

DOR Code  LEA Code Municipality      County  Cherry Sheet FY  Population  \
0           1         1     Abington    PLYMOUTH             2023       17062   
1           2         2        Acton   MIDDLESEX             2023       24021   
2           3         3     Acushnet     BRISTOL             2023       10559   
3           4         4        Adams   BERKSHIRE             2023        8166   
4           5         5       Agawam     HAMPDEN             2023       28692   
..        ...       ...          ...         ...              ...         ...   
346       347       347       Woburn   MIDDLESEX             2023       40876   
347       348       348    Worcester   WORCESTER             2023      206518   
348       349       349  Worthington   HAMPSHIRE             2023        1193   
349       350       350     Wrentham     NORFOLK             2023       12178   
350       351       351     Yarmouth  BARNSTABLE             2023       25023   

     DOR Income  DOR Income Per Capita          EQV  EQV Per Capita  
0     645765000                  37848   2379458500          139460  
1    1733168000                  72152   4940912200          205691  
2     376411000                  35648   1346443900          127516  
3     191139000                  23407    549332800           67271  
4     893064000                  31126   3286756800          114553  
..          ...                    ...          ...             ...  
346  1829233000                  44751   9226467200          225718  
347  4638107000                  22459  15426121200           74696  
348    34443000                  28871    182949300          153352  
349   738007000                  60602   2481220700          203746  
350   876692000                  35035   6880187700          274955  

[351 rows x 10 columns]

In [20]:
df_perpupil

District Name  District Code  \
0         Abby Kelley Foster Charter Public (District)        4450000   
1                                             Abington          10000   
2    Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                     Acton-Boxborough        6000000   
4                                             Acushnet          30000   
..                                                 ...            ...   
393                                             Woburn        3470000   
394                                          Worcester        3480000   
395                                        Worthington        3490000   
396                                           Wrentham        3500000   
397                                       State Totals              0   

     In-District Expenditures Total In-district FTEs  \
0              $24,231,438.78                1,415.8   
1              $35,603,421.16                2,143.5   
2              $12,801,946.92                  467.4   
3             $100,564,755.40                5,181.7   
4              $14,679,700.00                  965.8   
..                        ...                    ...   
393            $94,093,579.73                4,289.4   
394           $501,140,676.40               24,319.6   
395             $1,904,264.00                   75.5   
396            $17,626,509.66                  920.9   
397        $18,401,964,825.00              865,749.1   

    In-District Expenditures per Pupil        Total Expenditures  \
0                           $17,114.65            $24,231,438.78   
1                           $16,609.95            $41,098,932.87   
2                           $27,392.63            $12,801,946.92   
3                           $19,407.68           $109,746,603.20   
4                           $15,199.52            $18,752,410.80   
..                                 ...                       ...   
393                         $21,936.30           $105,328,116.50   
394                         $20,606.45           $568,546,854.00   
395                         $25,222.04             $2,822,565.38   
396                         $19,140.53            $19,059,388.49   
397                         $21,255.54        $20,737,958,663.00   

       Total Pupil FTEs Total Expenditures per Pupil  
0               1,415.8                   $17,114.65  
1               2,297.3                   $17,890.10  
2                 467.4                   $27,392.63  
3               5,302.1                   $20,698.70  
4               1,254.7                   $14,945.73  
..                  ...                          ...  
393             4,516.7                   $23,319.71  
394            27,234.8                   $20,875.75  
395               142.1                   $19,863.23  
396               981.1                   $19,426.55  
397           947,578.4                   $21,885.22  

[398 rows x 8 columns]

In [22]:
df_population

District Name  District Code  \
0         Abby Kelley Foster Charter Public (District)        4450000   
1                                             Abington          10000   
2    Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                     Acton-Boxborough        6000000   
4                                             Acushnet          30000   
..                                                 ...            ...   
394                                             Woburn        3470000   
395                                          Worcester        3480000   
396                                        Worthington        3490000   
397                                           Wrentham        3500000   
398                                       State Totals              0   

    High Needs#  High Needs% English Learners#  English Learners%  \
0         1,005         70.7               256               18.0   
1         1,029         47.3               236               11.0   
2           332         71.1                44                9.4   
3         1,538         29.6               317                6.2   
4           415         42.7                 6                0.6   
..          ...          ...               ...                ...   
394       2,332         53.6               461               10.8   
395      20,471         82.9             7,386               30.4   
396          41         53.9                 0                0.0   
397         284         29.8                24                2.5   
398     508,820         55.1           110,554               12.1   

    First Language Not English#  First Language Not English% Low Income#  \
0                           967                         68.0         837   
1                           362                         16.9         736   
2                           158                         33.8         273   
3                         1,141                         22.2         546   
4                            17                          1.8         320   
..                          ...                          ...         ...   
394                         894                         20.9       1,653   
395                      14,201                         58.4      17,636   
396                           0                          0.0          30   
397                          51                          5.4         135   
398                     228,169                         25.0     386,060   

     Low Income% Students with Disabilities# Students with Disabilities%  \
0           58.9                                                           
1           34.4                                                           
2           58.5                                                           
3           10.6                                                           
4           33.2                                                           
..           ...                         ...                         ...   
394         38.7                                                           
395         72.5                                                           
396         39.5                                                           
397         14.2                                                           
398         42.3                                                           

    Students with Disabilities#.1  Students with Disabilities%.1  
0                             166                           11.7  
1                             387                           17.8  
2                             115                           24.6  
3                             803                           15.5  
4                             154                           15.8  
..                            ...                            ...  
394                           896       

In [24]:
df_teacher

District Name  District Code  \
0         Abby Kelley Foster Charter Public (District)        4450000   
1                                             Abington          10000   
2    Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                     Acton-Boxborough        6000000   
4                                             Acushnet          30000   
..                                                 ...            ...   
394                                             Woburn        3470000   
395                                          Worcester        3480000   
396                                        Worthington        3490000   
397                                           Wrentham        3500000   
398                                       State Totals              0   

    Total # of Teachers (FTE)  % of Teachers Licensed Student / Teacher Ratio  \
0                       119.6                    99.2               11.9 to 1   
1                       153.8                    97.1               13.9 to 1   
2                        49.6                    65.6                9.4 to 1   
3                       395.4                   100.0               13.0 to 1   
4                        76.0                   100.0               12.7 to 1   
..                        ...                     ...                     ...   
394                     374.6                    99.2               11.4 to 1   
395                   1,715.9                    97.6               14.2 to 1   
396                       9.2                   100.0                8.3 to 1   
397                      79.4                   100.0               12.0 to 1   
398                  76,977.6                    97.1               11.9 to 1   

     Percent of Experienced Teachers  \
0                               81.5   
1                               81.3   
2                               37.3   
3                               88.3   
4                               78.9   
..                               ...   
394                             85.7   
395                             77.3   
396                             69.6   
397                             85.3   
398                             81.3   

     Percent of Teachers without Waiver or Provisional License  \
0                                                 88.3           
1                                                 89.6           
2                                                 95.6           
3                                                 97.5           
4                                                 92.1           
..                                                 ...           
394                                               96.1           
395                                               89.8           
396                                               87.0           
397                                              100.0           
398                                               93.4           

     Percent Teaching In-Field  
0                         90.5  
1                         95.7  
2                         47.3  
3                         98.8  
4                         94.7  
..                         ...  
394                       95.7  
395                       93.0  
396                      100.0  
397                      100.0  
398                       92.3  

[399 rows x 8 columns]

In [30]:
df_population.isnull().sum() 

## the original dataset doesn't have the columns "Students with Disabilities# , Students with Disabilities% " 
## For this reason I checked why there is no data in these columns. It doesn't say null but it doesn't have any info either. 
## After I combined the datasets, I dropped these columns.

District Name                    0
District Code                    0
High Needs#                      0
High Needs%                      0
English Learners#                0
English Learners%                0
First Language Not English#      0
First Language Not English%      0
Low Income#                      0
Low Income%                      0
Students with Disabilities#      0
Students with Disabilities%      0
Students with Disabilities#.1    0
Students with Disabilities%.1    0
dtype: int64

### Merge Datasets (LEFT join)

* I applied LEFT Join
* Keep the df_population datasets as left table since it has the highest number of rows(399)
* Then merged each datasets and dropped the "District Name" other wise it was creating duplicates

In [34]:
# Start with base DataFrame
df = df_population.copy()

# Merge and drop redundant 'District Name' after each merge
df = df.merge(df_acctblty.drop(columns=['District Name']), on='District Code', how='left')
df = df.merge(df_perpupil.drop(columns=['District Name']), on='District Code', how='left')
df = df.merge(df_grads.drop(columns=['District Name']), on='District Code', how='left')
df = df.merge(df_teacher.drop(columns=['District Name']), on='District Code', how='left')

In [42]:
df.shape

(399, 36)

In [44]:
df.columns 

Index(['District Name', 'District Code', 'High Needs#', 'High Needs%',
       'English Learners#', 'English Learners%', 'First Language Not English#',
       'First Language Not English%', 'Low Income#', 'Low Income%',
       'Students with Disabilities#', 'Students with Disabilities%',
       'Students with Disabilities#.1', 'Students with Disabilities%.1',
       'Overall Classification', 'Reason for Classification',
       'Progress Toward Improvement Targets (%)', 'In-District Expenditures',
       'Total In-district FTEs', 'In-District Expenditures per Pupil',
       'Total Expenditures', 'Total Pupil FTEs',
       'Total Expenditures per Pupil', '# in Cohort', '% Graduated',
       '% Still in School', '% Non-Grad Completers', '% H.S. Equiv.',
       '% Dropped Out', '% Permanently Excluded', 'Total # of Teachers (FTE)',
       '% of Teachers Licensed', 'Student / Teacher Ratio',
       'Percent of Experienced Teachers',
       'Percent of Teachers without Waiver or Provisional L

In [46]:
df.isnull().sum()

District Name                                                 0
District Code                                                 0
High Needs#                                                   0
High Needs%                                                   0
English Learners#                                             0
English Learners%                                             0
First Language Not English#                                   0
First Language Not English%                                   0
Low Income#                                                   0
Low Income%                                                   0
Students with Disabilities#                                   0
Students with Disabilities%                                   0
Students with Disabilities#.1                                 0
Students with Disabilities%.1                                 0
Overall Classification                                        1
Reason for Classification               

In [48]:
df.dtypes

District Name                                                 object
District Code                                                  int64
High Needs#                                                   object
High Needs%                                                  float64
English Learners#                                             object
English Learners%                                            float64
First Language Not English#                                   object
First Language Not English%                                  float64
Low Income#                                                   object
Low Income%                                                  float64
Students with Disabilities#                                   object
Students with Disabilities%                                   object
Students with Disabilities#.1                                 object
Students with Disabilities%.1                                float64
Overall Classification            

### Save the Combined Dataset

In [51]:
# save the final dataframe to a new Excel file
output_file = "combined_data.xlsx"
df.to_excel(output_file, index=False)

In [53]:
# print the first few rows of the final dataframe
pd.set_option('display.max_columns', None)
df.head()

District Name  District Code  \
0       Abby Kelley Foster Charter Public (District)        4450000   
1                                           Abington          10000   
2  Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                   Acton-Boxborough        6000000   
4                                           Acushnet          30000   

  High Needs#  High Needs% English Learners#  English Learners%  \
0       1,005         70.7               256               18.0   
1       1,029         47.3               236               11.0   
2         332         71.1                44                9.4   
3       1,538         29.6               317                6.2   
4         415         42.7                 6                0.6   

  First Language Not English#  First Language Not English% Low Income#  \
0                         967                         68.0         837   
1                         362                         16.9         736   
2                         158                         33.8         273   
3                       1,141                         22.2         546   
4                          17                          1.8         320   

   Low Income% Students with Disabilities# Students with Disabilities%  \
0         58.9                                                           
1         34.4                                                           
2         58.5                                                           
3         10.6                                                           
4         33.2                                                           

  Students with Disabilities#.1  Students with Disabilities%.1  \
0                           166                           11.7   
1                           387                           17.8   
2                           115                           24.6   
3                           803                           15.5   
4                           154                           15.8   

                     Overall Classification  \
0  Not requiring assistance or intervention   
1  Not requiring assistance or intervention   
2  Not requiring assistance or intervention   
3  Not requiring assistance or intervention   
4  Not requiring assistance or intervention   

             Reason for Classification  \
0         Meeting or exceeding targets   
1     Moderate progress toward targets   
2     Moderate progress toward targets   
3         Meeting or exceeding targets   
4  Substantial progress toward targets   

   Progress Toward Improvement Targets (%)  In-District Expenditures  \
0                                     82.0            $24,231,438.78   
1                                     39.0            $35,603,421.16   
2                                     48.0            $12,801,946.92   
3                                     85.0           $100,564,755.40   
4                                     58.0            $14,679,700.00   

  Total In-district FTEs In-District Expenditures per Pupil  \
0                1,415.8                         $17,114.65   
1                2,143.5                         $16,609.95   
2                  467.4                         $27,392.63   
3                5,181.7                         $19,407.68   
4                  965.8                         $15,199.52   

         Total Expenditures    Total Pupil FTEs Total Expenditures per Pupil  \
0            $24,231,438.78             1,415.8                   $17,114.65   
1            $41,098,932.87             2,297.3                   $17,890.10   
2            $12,801,946.92               467.4                   $27,392.63   
3           $109,746,603.20             5,302.1                   $20,698.70   
4            $18,752,410.80             1,254.7                   $14,945.73   

  # in Cohort  % Graduated  % Still in School  % Non-Grad Completers  \
0          88        

In [57]:
df = df.drop(columns=['Students with Disabilities%', 'Students with Disabilities#'])

In [59]:
df.shape

(399, 34)